src/trio/_core/_ki.py

```diff
@@ -83,20 +78,112 @@
 # for any Python program that's written to catch and ignore
 # KeyboardInterrupt.)
 
-# We use this special string as a unique key into the frame locals dictionary.
-# The @ ensures it is not a valid identifier and can't clash with any possible
-# real local name. See: https://github.com/python-trio/trio/issues/469
-LOCALS_KEY_KI_PROTECTION_ENABLED: Final = "@TRIO_KI_PROTECTION_ENABLED"
+_T = TypeVar("_T")
+
+
+class _IdRef(weakref.ref[_T]):
+    __slots__ = ("_hash",)
+    _hash: int
+
+    def __new__(cls, ob: _T, callback: Callable[[Self], Any] | None = None, /) -> Self:
+        self: Self = weakref.ref.__new__(cls, ob, callback)
+        self._hash = object.__hash__(ob)
+        return self
+
+    def __eq__(self, other: object) -> bool:
+        if self is other:
+            return True
+
+        if not isinstance(other, _IdRef):
+            return NotImplemented
+
+        my_obj = None
+        try:
+            my_obj = self()
+            return my_obj is not None and my_obj is other()
+        finally:
+            del my_obj
+
+    # we're overriding a builtin so we do need this
+    def __ne__(self, other: object) -> bool:
+        return not self == other
+
+    def __hash__(self) -> int:
+        return self._hash
+
+
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+
+
+# see also: https://github.com/python/cpython/issues/88306
+class WeakKeyIdentityDictionary(Generic[_KT, _VT]):
+    def __init__(self) -> None:
+        self._data: dict[_IdRef[_KT], _VT] = {}
+
+        def remove(
+            k: _IdRef[_KT],
+            selfref: weakref.ref[
+                WeakKeyIdentityDictionary[_KT, _VT]
+            ] = weakref.ref(  # noqa: B008  # function-call-in-default-argument
```

Comment:
Tangential: Is there a ruff setting or something where we can use long descriptive names instead of stuff like `B008`? I certainly don't like having to look them up.

Reply:
No there's not :(

Reply:
Not yet, but it is planned: https://github.com/astral-sh/ruff/issues/1773. Needs decisions on how it should work UI wise, etc.